pkg/controller/add_cluster.go

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package controller
+
+import (
+	"fmt"
+
+	plc "github.com/m3db/m3/src/query/api/v1/handler/placement"
+	"github.com/m3db/m3/src/query/generated/proto/admin"
+	"github.com/m3db/m3cluster/generated/proto/placementpb"
+	myspec "github.com/m3db/m3db-operator/pkg/apis/m3dboperator/v1"
+	"github.com/m3db/m3db-operator/pkg/m3admin"
+	"github.com/m3db/m3x/ident"
+
+	"go.uber.org/zap"
+)
+
+const (
+	// defaults for placement init request
+	_defaultM3DBPort = 9000
+)
+
+func (c *Controller) addM3DBCluster(cluster *myspec.M3DBCluster) error {
+
+	redStatus := myspec.M3DBStatus{
+		State:   myspec.RedState,
+		Message: "m3 db failed to initialize",
+	}
+
+	// Refresh cluster map with latest state
+	if err := c.refreshClusters(); err != nil {
+		c.logger.Error("error refreshing cluster", zap.Error(err))
+		return err
+	}
+	// Ensure local copies of the cluster object from the internal map
+	cls := c.clusters[cluster.GetName()]
+	status := myspec.M3DBStatus{
+		State:   myspec.YellowState,
+		Message: "m3 db is initializing",
+	}
+	if err := c.updateM3DBStatus(cls.M3DBCluster, status); err != nil {
+		return err
+	}
+	// create services defined in service configuration
+	for _, svcCfg := range cls.M3DBCluster.Spec.ServiceConfigurations {
+		if err := c.k8sclient.EnsureService(cls.M3DBCluster, svcCfg); err != nil {
+			redStatus.Message = fmt.Sprintf("%s: %v", redStatus.Message, err)
+			if err := c.updateM3DBStatus(cls.M3DBCluster, redStatus); err != nil {
+				return err
+			}
+			return err
+		}
+	}
+
+	// TODO(PS) remove loop to find the correct service
+	for _, svcCfg := range cls.M3DBCluster.Spec.ServiceConfigurations {
+		if svcCfg.Name == "m3dbnode" {
```

Comment:
hm maybe use a function to get a svc's config from the spec, it'd simplify the code a bunch, e.g. 

```
func getServiceConfig(svcConfigs ..., serviceName string) (config, bool) {
  for _, svcConfig := range svcConfigs { 
    if svcConfig.Name == serviceName {
      return svcConfig, true
    }
  }
  return config{}, false
}
```